README.md
# Follow the stock

It follows some stocks on [boursorama](http://www.boursorama.com) and send some alerts through XMPP when it detects some changes over the subscribed percentage threshold.

# Command line interface

Here are the command line interfaces arguments to start the app:

    usage: followthestock -config <file>
      -config="/etc/followthestock/followthestock.conf": Config file
      -console=false: Use console
      -show-config=false: Show config

# Config file

The config file looks something like that:

    [general]
    exactTiming = false
    
    [xmpp]
    username = <username>
    password = <password>
    server = talk.google.com:443
    notls = false
    debug = true
    activityWatchdogMinutes = 30

    [db]
    # In the current working directory (should be /var/lib/followthestock)
    file = followthestock.db

# Client comands

Each client can send the following commands:

* `!help` - Display help
* `!s <stock> <per>` - Subscribe to variation about a stock
* `!u <stock>` - Unsubscribe from a stock
* `!g <stock>` - Get data about a stock
* `!ls` - List currently monitored stocks
* `!v <stock> <nb> <cost>` - Register the cost of our current stocks to calculate the added value
* `!pause <days>` - Pause alerts for X days
* `!resume` - Resume alerts
* `!uptime` - Bot uptime

Here are valid stock formats:
* `RNO` is like `FR:RNO`, which is the french "RENAULT" stock
* `US:RNO` is the "RHINO RESOURCE PARTNERS LP" stock
* `FR0011574110` is like `W:FR0011574110` which is the "SOGEN 50C 0614S" warrant


# Stocks data source
The stocks are fetched from [boursorama](http://www.boursorama.com). It is not an official API, it might not be legal to fetch data and it might not work in the future.

# Debian packages
Debian packages are automatically generated here:
 http://94.23.55.152/followthestock/dist/package/


debian/changelog
followthestock (0.4.2) unstable; urgency=low

  * Fixed feedback loop

 -- Florent Clairambault <florent@xps>  Wed, 11 Mar 2015 22:55:46 +0100

followthestock (0.4.1) unstable; urgency=low

  * Negative condition alert fix

 -- Florent Clairambault <florent@xps>  Wed, 11 Mar 2015 22:11:02 +0100

followthestock (0.4-2) unstable; urgency=low

  * Updating log format

 -- Florent Clairambault <florent@xps>  Tue, 10 Mar 2015 23:19:05 +0100

followthestock (0.4-1) unstable; urgency=low

  * Removing the "!" before each command
  * Ignoring uppercase (for easier mobile phone control)
  * XMPP handling code cleanup

 -- Florent Clairambault <florent@xps>  Tue, 10 Mar 2015 21:23:34 +0100

followthestock (0.4) unstable; urgency=low

  * Added support for plus/minus percentage change ("!s rno -1%")
  * Added support for time-window alerts ("!s rno 1% 24h")
  * Added belgium market
  * Modified the database (automatic upgrade) to support new data
  * Modified the database to add indexes
  * Simplified some errors

 -- Florent Clairambault <florent@xps>  Tue, 10 Mar 2015 00:35:53 +0100

followthestock (0.3) unstable; urgency=low

  * Added URL support in the alerts (!url / !nourl)
  * Modified the database (automatic upgrade of existing database)
  * Small code cleanup

 -- Florent Clairambault <florent@xps>  Thu, 05 Mar 2015 00:48:58 +0100

followthestock (0.2-5) unstable; urgency=low

  * Fixed a regex
  * Simplified regex handling

 -- Florent Clairambault <florent@xps>  Sun, 01 Mar 2015 21:46:22 +0100

followthestock (0.2-4) unstable; urgency=low

  * Fixed failed fetches counter
  * Adding automatic deletion of unfetchable stocks

 -- Florent Clairambault <florent@xps>  Sun, 19 Oct 2014 13:39:27 +0200

followthestock (0.2-3) unstable; urgency=low

  * Bugfix: Saving stock when modified

 -- Florent Clairambault <florent@xps>  Sun, 19 Oct 2014 01:18:17 +0200

followthestock (0.2-2) unstable; urgency=low

  * Fixed named fetching REGEX
  * Fetching name and currency automatically when we don't have them

 -- Florent Clairambault <florent@xps>  Sun, 19 Oct 2014 00:57:11 +0200

followthestock (0.2-1) unstable; urgency=low

  * Added database upgrade support
  * Added simple db unit test

 -- Florent Clairambault <florent@xps>  Sat, 18 Oct 2014 16:58:59 +0200

followthestock (0.2) unstable; urgency=low

  * Adding US2 market
  * Fixing a bug with failed request creating a nil crash
  * Console is now a CLI option and disabled by default
  * Adding a count of failed fetches to remove currencies that are deleted

 -- Florent Clairambault <florent@xps>  Fri, 17 Oct 2014 00:13:25 +0200

followthestock (0.1-2) unstable; urgency=low

  * Allowing to change the watchdog activity timeout.

 -- Florent Clairambault <florent@xps>  Mon, 08 Sep 2014 21:52:27 +0200

followthestock (0.1-1) unstable; urgency=low

  * First packaging

 -- Florent Clairambault <florent@clairambault.fr>  Mon, 08 Sep 2014 22:27:37 +0200


db_test.go
package main

import (
	"testing"
)

func prepareConfig() {
	if config == nil {
		config = NewConfig()
	}
}

// It's never too late to add unit tests
func TestParameters(t *testing.T) {
	prepareConfig()
	db := NewFtsDB()
	defer db.Close()

	if err := db.SetParameter("test1", "value1"); err != nil {
		t.Fatal(err)
	}

	if value := db.GetParameter("test2"); value != nil {
		t.Fatalf(`value should be nil, it's "%s"`, value)
	}

	if value := db.GetParameter("test1"); value == nil || *value != "value1" {
		t.Fatalf(`value should be "test1", it's "%s"`, value)
	}

}

func TestStockDeletion(t *testing.T) {
	prepareConfig()
	db := NewFtsDB()
	defer db.Close()

	{ // Data insertion
		s := &Stock{
			Market: "FR",
			Short:  "RNO",
		}

		db.SaveStock(s)

		c := &Contact{
			Email: "florent@clairambault.fr",
		}

		db.SaveContact(c)

		a := &Alert{
			Stock:   s.Id,
			Contact: c.Id,
		}

		db.SaveAlert(a)
	}

	{ // Data check and deletion
		s := db.GetStock("FR", "RNO")
		if s == nil || s.Id == 0 {
			t.Fatalf("Wrong stock: %#v", s)
		}

		db.DeleteStock(s)
	}

	{ // Data deletion
		s := db.GetStock("FR", "RNO")
		if s != nil {
			t.Fatalf("We should not have a stock anymore: %#v", s)
		}
	}
}


db.go
package main

import (
	"database/sql"
	"errors"
	"fmt"
	"github.com/coopernurse/gorp"
	_ "github.com/mattn/go-sqlite3"
	"strconv"
	"strings"
	"time"
)

type Parameter struct {
	Name  string `db:"name"`
	Value string `db:"value"`
}

type Stock struct {
	// db tag lets you specify the column name if it differs from the struct field
	Id            int64   `db:"stock_id"`
	Market        string  `db:"market"` // "FR","US","US2",W","AM"
	Short         string  `db:"short"`
	Name          string  `db:"name"`
	Value         float32 `db:"value"` // Last value
	Currency      string  `db:"currency"`
	FailedFetches int64   `db:"failed_fetches"`
}

type CurrencyConversion struct {
	From       string  `db:"from"`
	To         string  `db:"to"`
	Rate       float32 `db:"rate"`
	LastUpdate int64   `db:"last_update"`
}

type ContactStockValue struct {
	Id      int64   `db:"stock_value_id"`
	Contact int64   `db:"contact_id"`
	Stock   int64   `db:"stock_id"`
	Nb      int32   `db:"nb"`
	Value   float32 `db:"value"`
}

type Contact struct {
	Id         int64  `db:"contact_id"`
	Email      string `db:"email"`
	PauseUntil int64  `db:"pause_until"`
	ShowUrl    bool   `db:"show_url"`
}

type Value struct {
	Id    int64   `db:"value_id"`
	Stock int64   `db:"stock_id"`
	Date  int64   `db:"date"`
	Value float32 `db:"value"`
}

type Alert struct {
	Id               int64   `db:"alert_id"`
	Contact          int64   `db:"contact_id"`
	Stock            int64   `db:"stock_id"`
	LastTriggered    int64   `db:"last_triggered"`
	LastValue        float32 `db:"last_value"`
	LastDate         int64   `db:"last_date"`
	Duration         int64   `db:"duration"`
	Percent          float32 `db:"percent"`
	PercentDirection int     `db:"percent_direction"`
}

const (
	ALERT_DIRECTION_BOTH = iota
	ALERT_DIRECTION_UP   = iota
	ALERT_DIRECTION_DOWN = iota
)

type DatabaseUpgrade struct {
	Version int
	Sql     []string
}

type FtsDB struct {
	connection *sql.DB
	mapping    *gorp.DbMap
}

const (
	TABLE_PARAMETER           = "parameter"
	TABLE_STOCK               = "stock"
	TABLE_CONTACT             = "contact"
	TABLE_VALUE               = "value"
	TABLE_ALERT               = "alert"
	TABLE_CONTACT_STOCK_VALUE = "contactstockvalue"
	TABLE_CURRENCY_CONVERSION = "currency_conversion"
)

func NewFtsDB() *FtsDB {
	// We connect to the database
	conn, err := sql.Open("sqlite3", config.Db.File)
	if err != nil {
		log.Fatal(err)
	}

	// We create the DbMap instance
	dbmap := &gorp.DbMap{Db: conn, Dialect: gorp.SqliteDialect{}}

	// We register the tables
	dbmap.AddTableWithName(Parameter{}, TABLE_PARAMETER).SetKeys(false, "Name")
	dbmap.AddTableWithName(Stock{}, TABLE_STOCK).SetKeys(true, "Id")
	dbmap.AddTableWithName(Contact{}, TABLE_CONTACT).SetKeys(true, "Id")
	dbmap.AddTableWithName(Value{}, TABLE_VALUE).SetKeys(true, "Id")
	dbmap.AddTableWithName(Alert{}, TABLE_ALERT).SetKeys(true, "Id")
	dbmap.AddTableWithName(CurrencyConversion{}, TABLE_CURRENCY_CONVERSION).SetUniqueTogether("from", "to")
	dbmap.AddTableWithName(ContactStockValue{}, TABLE_CONTACT_STOCK_VALUE).SetKeys(true, "Id")

	// We create the tables
	err = dbmap.CreateTablesIfNotExists()
	if err != nil {
		log.Fatal(err)
	}

	{ // WAL is faster & safer
		_, err = conn.Exec("pragma journal_mode = wal")
		if err != nil {
			log.Fatal(err)
		}
	}

	db := &FtsDB{connection: conn, mapping: dbmap}

	db.Upgrade()

	return db
}

// Performs an automatic database upgrade
func (db *FtsDB) Upgrade() {
	upgrades := []*DatabaseUpgrade{
		&DatabaseUpgrade{
			Version: 1,
			Sql: []string{
				`alter table ` + TABLE_STOCK + ` add column "failed_fetches" integer default 0`,
			},
		},
		&DatabaseUpgrade{
			Version: 2,
			Sql: []string{
				`alter table ` + TABLE_CONTACT + ` add column "show_url" integer default 1`,
			},
		},
		&DatabaseUpgrade{
			Version: 3,
			Sql: []string{
				`alter table ` + TABLE_ALERT + ` add column "percent_direction" integer default 0`,
				`alter table ` + TABLE_ALERT + ` add column "last_date" integer default 0`,
				`alter table ` + TABLE_ALERT + ` add column "duration" integer default 0`,
				`create index value_stock_date on ` + TABLE_VALUE + `(stock_id, date);`,
				`create index alert_stock on ` + TABLE_ALERT + `(stock_id);`,
			},
		},
	}

	// We get the current version
	version := 0
	if sVersion := db.GetParameter("db_version"); sVersion != nil {
		version, _ = strconv.Atoi(*sVersion)
	}

	// We perform automatic upgrades
	for _, up := range upgrades {
		if version < up.Version {
			version = up.Version
			for _, sql := range up.Sql {
				log.Warning(`Performing SQL upgrade... "%s"`, sql)
				if _, err := db.connection.Exec(sql); err != nil {
					log.Error(`Failed to apply query "%s" with error: %s`, sql, err)
				} else {
					log.Info("OK !")
				}
			}
			// We want to save the version as soon as possible.
			db.SetParameter("db_version", fmt.Sprintf("%d", version))
		}
	}
}

func (db FtsDB) Close() {
	if err := db.connection.Close(); err != nil {
		log.Info("connection.Close(): %v", err)
	}
}

func (db *FtsDB) GetContactFromEmail(email string) *Contact {
	// We remove the part after the "/"
	email = strings.SplitN(email, "/", 2)[0]

	c := &Contact{}
	err := db.mapping.SelectOne(c, "select * from "+TABLE_CONTACT+" where email=?", email)
	if err != nil {
		log.Warning("Creating contact ", email)
		c.Email = email
		err := db.mapping.Insert(c)
		if err != nil {
			log.Error("Could not insert:", err)
		}
	}

	return c
}

func (db *FtsDB) GetContactFromId(id int64) *Contact {
	c := &Contact{}
	err := db.mapping.SelectOne(c, "select * from "+TABLE_CONTACT+" where contact_id=?", id)
	if err != nil {
		c = nil
	}
	return c
}

func (db *FtsDB) DeleteContact(c *Contact) (err error) {
	_, err = db.mapping.Delete(c)
	return err
}

func (db *FtsDB) DeleteStock(s *Stock) error {
	for _, a := range *db.GetAlertsForStock(s) {
		db.DeleteAlert(&a)
	}
	_, err := db.mapping.Delete(s)
	return err
}

func (db *FtsDB) GetStock(market, short string) *Stock {
	//log.Printf("GetStock( \"%s\", \"%s\" );", market, short)
	s := &Stock{}
	err := db.mapping.SelectOne(s, "select * from "+TABLE_STOCK+" where market=? and short=?", market, short)

	if err != nil {
		return nil
	} else {
		return s
	}
}

func (db *FtsDB) GetStockFromId(id int64) *Stock {
	s := &Stock{}
	err := db.mapping.SelectOne(s, "select * from "+TABLE_STOCK+" where stock_id=?", id)
	if err != nil {
		return nil
	} else {
		return s
	}
}

func (db *FtsDB) GetAlertsForStock(s *Stock) *[]Alert {
	var alerts []Alert
	db.mapping.Select(&alerts, "select * from "+TABLE_ALERT+" where stock_id=?", s.Id)
	return &alerts
}

func (db *FtsDB) GetAllStocks() *[]Stock {
	var stocks []Stock
	db.mapping.Select(&stocks, "select * from "+TABLE_STOCK)
	return &stocks
}

func (db *FtsDB) SaveStockValue(stock *Stock, value float32, date int64) (err error) {
	if stock.Value != value {
		stock.Value = value
		err = db.SaveStock(stock)
	} else {
		err = nil
	}

	if err == nil {
		err = db.mapping.Insert(&Value{Stock: stock.Id, Date: date, Value: value})
	}

	return err
}

func (this *FtsDB) GetStockValue(stock *Stock, date int64) (*Value, error) {
	value := &Value{}
	err := db.mapping.SelectOne(value, "select * from "+TABLE_VALUE+" where stock_id=? and date>? order by date desc limit 1;", stock.Id, date)

	if err != nil {
		return nil, err
	} else {
		return value, nil
	}
}

func (db *FtsDB) SubscribeAlert(s *Stock, c *Contact, per float32, direction int, duration int64) (alert *Alert, err error) {
	_, err = db.UnsubscribeAlert(s, c)

	if err != nil {
		return nil, err
	}

	alert = &Alert{Stock: s.Id, Contact: c.Id, Percent: per, PercentDirection: direction, Duration: duration}

	err = db.SaveAlert(alert)

	return
}

func (db *FtsDB) UnsubscribeAlert(s *Stock, c *Contact) (ok bool, err error) {
	_, err = db.mapping.Exec("delete from "+TABLE_ALERT+" where stock_id=? and contact_id=?", s.Id, c.Id)
	return
}

func (db *FtsDB) SaveContact(c *Contact) (err error) {
	if c.Id != 0 {
		_, err = db.mapping.Update(c)
	} else {
		err = errors.New("Contact doesn't exist !")
	}
	return
}

func (db *FtsDB) SaveAlert(a *Alert) (err error) {
	if a.Id != 0 {
		_, err = db.mapping.Update(a)
	} else {
		err = db.mapping.Insert(a)
	}
	return
}

func (db *FtsDB) GetAlertsForContact(c *Contact) *[]Alert {
	var alerts []Alert
	db.mapping.Select(&alerts, "select * from "+TABLE_ALERT+" where contact_id=?", c.Id)
	return &alerts
}

func (db *FtsDB) DeleteAlert(a *Alert) (err error) {
	_, err = db.mapping.Delete(a)
	return
}

func (db *FtsDB) SaveStock(s *Stock) error {
	if s.Id != 0 {
		_, e := db.mapping.Update(s)
		return e
	} else {
		return db.mapping.Insert(s)
	}
}

func (db *FtsDB) GetContactStockValue(contactId, stockId int64) *ContactStockValue {
	csv := &ContactStockValue{Contact: contactId, Stock: stockId}
	db.mapping.SelectOne(csv, "select * from "+TABLE_CONTACT_STOCK_VALUE+" where contact_id=? and stock_id=?", csv.Contact, csv.Stock)
	return csv
}

func (db *FtsDB) GetContactStockValuesFromContact(c *Contact) *[]ContactStockValue {
	var values []ContactStockValue
	db.mapping.Select(&values, "select * from "+TABLE_CONTACT_STOCK_VALUE+" where contact_id=?", c.Id)
	return &values
}

func (db *FtsDB) SaveContactStockValue(csv *ContactStockValue) error {
	if csv.Id != 0 {
		_, err := db.mapping.Update(csv)
		return err
	} else {
		return db.mapping.Insert(csv)
	}

}

func (db *FtsDB) DeleteContactStockValue(s *ContactStockValue) (err error) {
	_, err = db.mapping.Delete(s)
	return
}

func (db *FtsDB) GetCurrencyConversion(from, to string) *CurrencyConversion {
	c := &CurrencyConversion{}
	err := db.mapping.SelectOne(c, "select * from "+TABLE_CURRENCY_CONVERSION+" where from=? and to=?", from, to)
	if err == nil {
		return c
	} else {
		return nil
	}
}

func (db *FtsDB) GetParameter(name string) *string {
	var value string
	if err := db.mapping.SelectOne(&value, "select value from "+TABLE_PARAMETER+" where name = ?", name); err == nil {
		return &value
	} else {
		return nil
	}
}

func (db *FtsDB) SetParameter(name, value string) error {
	p := &Parameter{Name: name, Value: value}
	if nb, err := db.mapping.Update(p); err == nil {
		if nb == 0 {
			err = db.mapping.Insert(p)
		}
		return err
	} else {
		return err
	}
}

func (db *FtsDB) SaveCurrencyConversion(c *CurrencyConversion) error {
	return db.mapping.Insert(c)
}

func (db *FtsDB) DeleteCurrencyConversion(c *CurrencyConversion) (err error) {
	_, err = db.mapping.Delete(c)
	return
}

func (csv *ContactStockValue) Exists() bool {
	return csv.Id != 0
}

func (s *Stock) String() string {
	return fmt.Sprintf("\"%s\" (%s:%s)", s.Name, s.Market, s.Short)
}

func (this *Alert) String() string {
	// NOTE: This performs a request on each String call
	stock := db.GetStockFromId(this.Stock)
	var direction string
	switch this.PercentDirection {
	case ALERT_DIRECTION_UP:
		direction = "+"
	case ALERT_DIRECTION_DOWN:
		direction = "-"
	default:
		direction = "~"
	}

	str := fmt.Sprintf("%s %s%.2f%%", stock.String(), direction, this.Percent)
	if this.Duration != 0 {
		str += fmt.Sprintf(" on %s", time.Duration(this.Duration))
	}
	str += fmt.Sprintf(" [%d]", this.Id)
	return str
}


main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var xm *FtsXmpp
var db *FtsDB
var stocks *StocksMgmt

var config *Config

var waitForRc chan int

const FTS_VERSION = "0.4"

func init() {
	waitForRc = make(chan int)
}

func console_handling() {
	in := bufio.NewReader(os.Stdin)
	for {
		fmt.Printf("> ")
		line, err := in.ReadString('\n')
		if err != nil {
			log.Fatal(err)
			continue
		}
		line = strings.TrimRight(line, "\n")

		tokens := strings.SplitN(line, " ", 2)
		if tokens[0] == "" {
			continue
		} else if tokens[0] == "quit" {
			waitForRc <- 0
		} else {
			fmt.Printf("\"%s\" not understood !\n", tokens[0])
		}
	}
}

func core() (rc int) {
	// We open the database
	db = NewFtsDB()
	defer db.Close()

	// We start the XMPP handling code
	xm = NewFtsXmpp()
	xm.Start()
	defer xm.Stop()

	// We load the stocks
	stocks = NewStocksMgmt()
	stocks.Start()
	defer stocks.Stop()

	if Console {
		// We block on the console handling code
		go console_handling()
	}

	// We wait for someone to trigger the result code
	rc = <-waitForRc

	log.Info("Stopping !")

	db.Close()

	return
}

func main() {

	log.Info("Starting !")

	// We parse the parameters
	config = NewConfig()

	rc := core()

	log.Info("Bye !")

	os.Exit(rc)
}


stocks.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"math"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

type StockFollower struct {
	Stock *Stock
}

var (
	reCotation *regexp.Regexp
	reName     []*regexp.Regexp
	sleepTime  time.Duration = time.Minute
)

const TEMPDIR = "/tmp/followthestock/"

func init() {
	reCotation = regexp.MustCompile("<span class=\"cotation\">([0-9\\ \\.]+)[^A-Z<>]*([A-Z]{2,3})</span>")

	reName = []*regexp.Regexp{
		regexp.MustCompile("(?s)<[^>]* itemprop=\"name\" title=\"([^\\\"]+)\"[^>]*>"),
		regexp.MustCompile("(?s)<h1>.*<a.*>(.*)</a>.*</h1>"),
	}
}

func NewStockFollower(s *Stock) *StockFollower {
	return &StockFollower{Stock: s}
}

func (sf *StockFollower) run() {
	t := time.Now().UTC() //.UnixNano()
	for {
		v, _, err := sf.Stock.GetValue()
		if err != nil {
			log.Warning("Stock %s: %v", sf.Stock.String(), err)
		} else {
			log.Info("Stock %s = %f %s", sf.Stock, v, sf.Stock.Currency)
			sf.considerValue(v)
		}
		if config.General.ExactTiming {
			t = t.Add(sleepTime) //.Nanoseconds()
			sl := t.Sub(time.Now().UTC())
			time.Sleep(sl)
		} else {
			time.Sleep(sleepTime)
		}
	}
}

func (sf *StockFollower) considerValue(value float32) {

	now := time.Now().UTC().UnixNano()

	if value == 0 {
		log.Warning("We have to ignore zero value for stock %v.", sf.Stock)
		return
	}

	db.SaveStockValue(sf.Stock, value, now)
	for _, al := range *db.GetAlertsForStock(sf.Stock) {
		if al.LastValue == 0 {
			value = value * 0.5
			al.LastValue = value
			al.LastTriggered = now
			db.SaveAlert(&al)

			contact := db.GetContactFromId(al.Contact)
			if contact == nil {
				log.Info("Alert %d - Contact missing, deleting alert !", al.Id)
				db.DeleteAlert(&al)
			}

			continue
		}

		diff := value - al.LastValue
		per := diff / al.LastValue * 100
		varPer := float32(math.Abs(float64(per)))
		log.Info("Alert %s / %1.2f%%", al.String(), per)

		var triggered bool
		switch al.PercentDirection {
		case ALERT_DIRECTION_BOTH:
			triggered = (varPer >= al.Percent)
		case ALERT_DIRECTION_UP:
			triggered = (per > al.Percent)
		case ALERT_DIRECTION_DOWN:
			triggered = (per < -al.Percent)
		}

		if triggered {
			contact := db.GetContactFromId(al.Contact)
			if contact == nil {
				log.Info("Alert %d - Contact missing, deleting alert !", al.Id)
				db.DeleteAlert(&al)
				continue
			}
			if now < contact.PauseUntil {
				log.Info("Alert %d - Contact is in pause", al.Id)
				continue
			}

			log.Info("Alert %d - Trigger !", al.Id)
			al.LastValue = value
			timeDiff := time.Duration(now - al.LastTriggered)
			timeDiff -= timeDiff % time.Second
			al.LastTriggered = now
			al.LastDate = now
			message := fmt.Sprintf("%s : %.3f (%+.2f%%) in %v", sf.Stock.String(), value, per, timeDiff)

			if contact.ShowUrl {
				message += " / " + sf.Stock.Url()
			}

			db.SaveAlert(&al)

			// We might be able to give some valuation data
			if csv := db.GetContactStockValue(al.Contact, al.Stock); csv.Exists() {
				cost := float32(csv.Nb) * csv.Value
				value := float32(csv.Nb) * value
				diff := value - cost
				per := diff / cost * 100
				message += fmt.Sprintf(" / %.3f - %.3f = %+.3f (%+.2f%%)", value, cost, diff, per)
			}

			xm.Send <- &SendChat{Remote: contact.Email, Text: message}
		} else {
			// If we have a duration, we might have to push the LastDate in the future
			if al.Duration != 0 && now-al.LastDate > al.Duration {
				startOfTimeWindow := now - al.Duration
				if value, err := db.GetStockValue(sf.Stock, startOfTimeWindow); err == nil {
					al.LastDate = startOfTimeWindow
					al.LastValue = value.Value
					db.SaveAlert(&al)
					log.Info("Alert %s: lastDate = %v, lastValue = %v", al.String(), time.Unix(0, al.LastDate), al.LastValue)
				} else {
					log.Error("Cannot find rows for alert %v: %v", sf.Stock, err)
				}
			}
		}
	}
}

func (sf *StockFollower) Start() {
	go sf.run()
}

func (sf *StockFollower) Stop() {

}

func (sf *StockFollower) String() string {
	return sf.Stock.String()
}

type StocksMgmt struct {
	sync.RWMutex
	stocks map[string]*StockFollower
}

func httpGet(url string) (*http.Response, error) {
	log.Debug("Fetching \"%s\"...", url)
	r, e := http.Get(url)
	//log.Println("Fetched ", url)
	return r, e
}

func (this *Stock) PageContent() (body string, err error) {
	resp, err := httpGet(this.Url())
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", errors.New(fmt.Sprintf("Wrong status code %d", resp.StatusCode))
	}

	finalUrl := resp.Request.URL.String()

	if strings.Contains(finalUrl, "recherche") {
		return "", errors.New(fmt.Sprintf("Not found !"))
	}

	{ // We get the body
		raw, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			return "", err
		}
		body = string(raw)
	}

	return
}

func tryNewStock(market, short string) (*Stock, error) {
	log.Debug("tryNewStock( \"%s\", \"%s\" );", market, short)
	s := &Stock{Market: market, Short: short}

	body, err := s.PageContent()

	if err != nil {
		return nil, errors.New(fmt.Sprintf("No \"%s\" on %s market !", short, market))
	}

	for _, re := range reName { // Second attempt for other quotations
		result := re.FindStringSubmatch(body)
		if len(result) > 1 {
			s.Name = strings.Trim(result[1], " \n\r")
			break
		} else {
			log.Error("Regex failed: %s", re)
		}
	}

	if len(s.Name) == 0 { // If we still couldn't get a name
		// We will save the raw data for future testing
		os.MkdirAll(TEMPDIR, 0755)
		fileName := fmt.Sprintf("%s/%s_%s.html", TEMPDIR, s.Market, s.Short)
		if err := ioutil.WriteFile(fileName, []byte(body), 0644); err != nil {
			log.Error("Could not write %s: %s", fileName, err)
		}
		return s, errors.New("Could not get the name")
	}

	return s, nil
}

var marketsToTest = [...]string{"FR", "AM", "US", "US2", "W", "BE"}

func (s *Stock) boursoramaSymbol() (symbol string) {
	switch s.Market {
	case "US": // NASDAQ & NYSE
		symbol = s.Short
	case "US2": // XETRA ?
		symbol = "1z" + s.Short
	case "FR": // EURONEXT Paris
		symbol = "1rP" + s.Short
	case "AM": // EURONEXT Amsterdam
		symbol = "1rA" + s.Short
	case "W": // Warrants
		symbol = "2rP" + s.Short
	case "W2":
		symbol = "3rP" + s.Short
	case "BE": // EURONEXT Bruxelles
		symbol = "FF11-" + s.Short
	default:
		log.Fatal("Unknown market: " + s.Market)
	}
	return
}

func (this *Stock) Url() string {
	return fmt.Sprintf("http://www.boursorama.com/cours.phtml?symbole=%s", this.boursoramaSymbol())
}

func (s *Stock) fetchPage() (string, error) {
	resp, err := httpGet(fmt.Sprintf("http://www.boursorama.com/cours.phtml?symbole=%s", s.boursoramaSymbol()))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", errors.New(fmt.Sprintf("Wrong status code %d", resp.StatusCode))
	}

	finalUrl := resp.Request.URL.String()

	if strings.Contains(finalUrl, "recherche") {
		return "", errors.New(fmt.Sprintf("Not found !"))
	}

	{
		raw, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			return "", err
		}
		body := string(raw)
		return body, nil
	}
}

func (s *Stock) GetValue() (value float32, currency string, err error) {
	body, err := s.fetchPage()

	save := false

	result := reCotation.FindStringSubmatch(body)
	if len(result) >= 2 {
		v, _ := strconv.ParseFloat(strings.Replace(result[1], " ", "", -1), 32)
		value = float32(v)
		currency = result[2]
		if s.FailedFetches != 0 {
			s.FailedFetches = 0
			log.Debug("Updating %v's failed fetches", s)
			save = true
		}
		if s.Currency == "" && currency != "" {
			s.Currency = currency
			log.Info("Updating %v's currency", s)
			save = true
		}
	} else {
		s.FailedFetches += 1
		log.Warning("Could not fetch cotation %s for the %dth time.", s, s.FailedFetches)
		if s.FailedFetches > 1000 {
			log.Info("Deleting stock %#v ...", s)
			db.DeleteStock(s)
		} else {
			save = true
		}

	}

	if s.Name == "" || s.Currency == "" { // We get the name if we couldn't get it earlier
		log.Warning("Missing name or currency, trying to update data about %v", s)
		if s2, err := tryNewStock(s.Market, s.Short); err == nil {
			if s.Name == "" && s2.Name != "" {
				log.Info("Updating %v's name", s)
				s.Name = s2.Name
				save = true
			}
		}
	}

	if save {
		log.Info("Updating stock %#v ...", s)
		db.SaveStock(s)
	}

	return
}

func NewStocksMgmt() *StocksMgmt {
	sm := &StocksMgmt{stocks: make(map[string]*StockFollower)}

	return sm
}

func (sm *StocksMgmt) getOrCreateStock(market, short string) (s *Stock, e error) {
	s = db.GetStock(market, short)
	if s == nil { // If we couldn't get it
		s, e = tryNewStock(market, short) // We try to get it
		if s != nil {
			s.Value, s.Currency, e = s.GetValue() // And we get the value
			db.SaveStock(s)
		}
	} else if s.Currency == "" {
		_, s.Currency, _ = s.GetValue()
		db.SaveStock(s)
	}
	return
}

func (sm *StocksMgmt) GetStock(short string) (s *Stock, e error) {
	short = strings.ToUpper(short)
	tokens := strings.SplitN(short, ":", 2)

	if len(tokens) == 2 { // Specific market stock
		market := tokens[0]
		short = tokens[1]
		s, e = sm.getOrCreateStock(market, short)
	} else { // Unspecified market stock
		for _, market := range marketsToTest { // We test all stocks
			s, e = sm.getOrCreateStock(market, short)
			if s != nil {
				break
			}
		}
	}

	return
}

func (sm *StocksMgmt) LoadStock(s *Stock) {
	sf := NewStockFollower(s)
	sf.Start()
	sm.stocks[s.String()] = sf
}

func (sm *StocksMgmt) LoadStocks() (err error) {
	stocks := db.GetAllStocks()

	for _, s := range *stocks {
		log.Info("Loading %s...", s.String())
		stock := s // Not doing so make us share the same pointer
		sm.LoadStock(&stock)
	}

	return nil
}

func (sm *StocksMgmt) SubscribeAlert(s *Stock, c *Contact, per float32, direction int, duration int64) (alert *Alert, err error) {
	a, e := db.SubscribeAlert(s, c, per, direction, duration)

	sm.Lock()
	if _, ok := sm.stocks[s.String()]; !ok {
		sm.LoadStock(s)
	}
	sm.Unlock()

	return a, e
}

func (sm *StocksMgmt) UnsubscribeAlert(s *Stock, c *Contact) (err error) {
	_, err = db.UnsubscribeAlert(s, c)

	return
}

func (sm *StocksMgmt) Start() {
	sm.Lock()
	sm.LoadStocks()
	sm.Unlock()
}

func (sm *StocksMgmt) Stop() {
	log.Debug("StocksMgmt.Stop()")
}


config.go
package main

import (
	"code.google.com/p/gcfg"
	"flag"
	"fmt"
	"os"
)

type Config struct {
	Xmpp struct {
		Server                  string
		Username                string
		Password                string
		Notls                   bool
		Debug                   bool
		LinesPerMessage         int
		ActivityWatchdogMinutes int
	}

	General struct {
		ExactTiming bool
	}

	Db struct {
		File string
	}
}

var Console bool

func NewConfig() *Config {
	config := &Config{}
	config.Db.File = "followthestock.db"
	config.Xmpp.Username = ""
	config.Xmpp.Server = "talk.google.com:443"
	config.Xmpp.LinesPerMessage = 15
	config.Xmpp.ActivityWatchdogMinutes = 30

	var fileName string
	var showConfig bool
	flag.StringVar(&fileName, "config", "/etc/followthestock/followthestock.conf", "Config file")
	flag.BoolVar(&showConfig, "show-config", false, "Show config")
	flag.BoolVar(&Console, "console", false, "Use console")

	flag.Usage = func() {
		fmt.Fprintln(os.Stderr, "usage: followthestock -config <file>")
		flag.PrintDefaults()
		os.Exit(2)
	}
	flag.Parse()
	if err := gcfg.ReadFileInto(config, fileName); err != nil {
		log.Fatalf("Could not read config: %v", err)
	}

	if showConfig {
		log.Info("Config: %#v", config)
	}

	return config
}
